specs/experimental/custom-gas-token.md

```diff
@@ -0,0 +1,391 @@
+# Custom Gas Token
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**
+
+- [Overview](#overview)
+- [Constants](#constants)
+- [Properties of a Gas Paying Token](#properties-of-a-gas-paying-token)
+  - [Automated Validation](#automated-validation)
+- [Configuring the Gas Paying Token](#configuring-the-gas-paying-token)
+- [Contract Modifications](#contract-modifications)
+  - [OptimismPortal](#optimismportal)
+    - [`depositERC20Transaction`](#depositerc20transaction)
+      - [Function Arguments](#function-arguments)
+    - [`depositTransaction`](#deposittransaction)
+    - [`setGasPayingToken`](#setgaspayingtoken)
+  - [StandardBridge](#standardbridge)
+  - [CrossDomainMessenger](#crossdomainmessenger)
+  - [SystemConfig](#systemconfig)
+  - [L1Block](#l1block)
+- [User Flow](#user-flow)
+- [Upgrade](#upgrade)
+  - [L1Block Deployment](#l1block-deployment)
+  - [L2CrossDomainMessenger Deployment](#l2crossdomainmessenger-deployment)
+  - [L2StandardBridge Deployment](#l2standardbridge-deployment)
+  - [L1Block Proxy Update](#l1block-proxy-update)
+  - [L2CrossDomainMessenger Proxy Update](#l2crossdomainmessenger-proxy-update)
+  - [L2StandardBridge Proxy Update](#l2standardbridge-proxy-update)
+- [Selection of `ETHER_TOKEN_ADDRESS`](#selection-of-ether_token_address)
+- [Standard Config](#standard-config)
+- [Security Considerations](#security-considerations)
+  - [OptimismPortal Token Allowance](#optimismportal-token-allowance)
+  - [Decimal Scaling](#decimal-scaling)
+  - [Interoperability Support](#interoperability-support)
+  - [Wrapped Ether](#wrapped-ether)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Overview
+
+Custom gas token is also known as "custom L2 native asset". It allows for an L1-native ERC20 token to collateralize
+and act as the gas token on L2. The native asset is used to buy gas which is used to pay for resources on the network
+and is represented by EVM opcodes such as `CALLVALUE` (`msg.value`).
+
+Both the L1 and L2 smart contract systems MUST be able to introspect on the address of the gas paying token to
+ensure the security of the system. The source of truth for the configuration of the address of the token is
+in the L1 `SystemConfig` smart contract.
+
+The terms "custom gas token", "gas paying token", "native asset" and "gas paying asset" can all be used interchangebly.
+Ether is not a token.
+
+## Constants
+
+| Constant          | Value                           | Description |
+| ----------------- | ------------------------------- | ----------- |
+| `ETHER_TOKEN_ADDRESS` |  `address(0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE)` | Represents ether for gas paying asset |
```

Comment:
`ETHER_TOKEN_ADDRESS` could be misunderstood to mean a token related to eth (i.e., a tokenized instance of ETH like WETH). could `BASE_TOKEN_ADDRESS` or `GAS_TOKEN_ADDRESS` make more sense?